### PR TITLE
Export traceability information as `SystemTask`

### LINKED ISSUE #4010: Improve traceability migration logic
### Problem

Though possible it was preferred to have more light weight operations with cleaner separations.
Traceability migration was turned off and reverted for 0.30 and [in current 0.31](https://github.com/hashgraph/hedera-services/pull/4008).
A more light weight operation with cleaner separations and checks was preferred.

### Solution

Improve the traceability migration logic by
- utilize an explicitly static property to enable migration `hedera.recordStream.enableTraceabilityMigration`
- carry out state migration across multiple handleTransactions to reduce the size of side cars for a given timestamp
- ensure migration occurs only once on upgrade and not a 2nd time on restart. Potentially utilize version comparisons to ensure only 0.30 -> 0.31 migration is applicable

### Alternatives

_No response_

### LINKED ISSUE #4011: Generalize expiry work as a `SystemTask` facility
### Problem

We have various types of work with the general profile that they,
 - Need the opportunity to run for, possibly, every entity in state; and,
 - Must happen deterministically, because they create records/change state; and,
 - Can have their performance impact measured in terms of `MerkleMap`/`VirtualMap` operations; and,
 - May either finish at some point, or run continuously.

### Solution

Generalize the facility used for expiration work to perform any implementation of a `SystemTask`.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Extracts the one-shot traceability migration into a [`TraceabilityExportTask`](https://github.com/hashgraph/hedera-services/blob/04011-export-traceability-via-system-task/hedera-node/src/main/java/com/hedera/services/state/tasks/TraceabilityExportTask.java#L74) that is (eventually) given the opportunity to export sidecars for every contract by [the `SystemTaskManager`](https://github.com/hashgraph/hedera-services/blob/04011-export-traceability-via-system-task/hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java).
 - Adds four new fields---`nextTaskTodo`, `seqNoPostUpgrade`,`lastScannedPostUpgrade`, and `preExistingEntityScanStatus`---to the `MerkleNetworkContext` for managing system tasks across restarts and reconnects.
     * These fields let us detect when all pre-existing entities have been scanned after an upgrade; and to skip exporting traceability info for contracts created _after_ the `0.31.x` upgrade. 
- Uses a system transaction record of a no-op synthetic `ContractUpdate` to export traceability sidecars.
- _Aside:_ Tweaks the `SteadyStateThrottlingCheck` to allow easily validating TPS of a target network, example [here](https://github.com/hashgraph/hedera-services/blob/04011-export-traceability-via-system-task/test-clients/src/main/java/com/hedera/services/bdd/suites/regression/SteadyStateThrottlingCheck.java#L158).

**Related issues**:
 - #4011 
 Fixes #4010

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryRecordsHelper.java [hunk 1]
```diff
@@ -108,19 +111,29 @@ public void streamCryptoRenewal(
                         .setHbarAdjustments(sideEffectsTracker.getNetTrackedHbarChanges())
                         .setStakingRewardsPaid(sideEffectsTracker.getStakingRewardsPaid())
                         .setFee(fee);
-        stream(expirableTxnRecord, synthBody, eventTime);
+        finalizeAndStream(expirableTxnRecord, synthBody, eventTime, recordStreaming);
     }
 
-    private void stream(
+    public static void finalizeAndStream(
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryRecordsHelper.java:117]: Probably we can add `StreamingUtils` for these static methods ?

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java [hunk 2]
```diff
@@ -98,10 +104,14 @@ public class MerkleNetworkContext extends PartialMerkleLeaf implements MerkleLea
...
     @Nullable private Instant consensusTimeOfLastHandledTxn = null;
     private SequenceNumber seqNo;
+    private int nextTaskTodo;
     private long lastScannedEntity;
+    private long seqNoPostUpgrade;
+    private long lastScannedPostUpgrade;
     private long entitiesScannedThisSecond = 0L;
     private long entitiesTouchedThisSecond = 0L;
     private long preparedUpdateFileNum = NO_PREPARED_UPDATE_FILE_NUM;
+    private byte preExistingEntityScanStatus;
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java:114]: Will be nice to have Javadoc for the new variables.

File: hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     * reached.
+     *
+     * <p>More precisely, this method will return:
+     *
+     * <ul>
+     *   <li>{@link SystemTaskResult#NOTHING_TO_DO} iff all tasks returned {@code NOTHING_TO_DO}.
+     *   <li>{@link SystemTaskResult#DONE} iff at least one task returned {@code DONE} and all other
+     *       tasks returned {@code NOTHING_TO_DO}.
+     *   <li>{@link SystemTaskResult#NO_CAPACITY_LEFT} immediately when any task reports it.
+     *   <li>{@link SystemTaskResult#NEEDS_DIFFERENT_CONTEXT} immediately when any task reports it.
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java:60]: Currently `NEEDS_DIFFERENT_CONTEXT` is returned only when the preceding and following child records exceed 54 ?

File: hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            }
+        }
+        curNetworkCtx.setNextTaskTodo(0);
+        // Only update the pre-existing scan status when we complete all tasks for an entity id
+        updatePreExistingScanStatus(literalNum, curNetworkCtx);
+        return didSomething ? DONE : NOTHING_TO_DO;
+    }
+
+    @VisibleForTesting
+    void updatePreExistingScanStatus(final long n, final MerkleNetworkContext curNetworkCtx) {
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java:90]: I thought sonar will warn if we have any method being used in production code with `@VisibleForTesting` 🤔

File: hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskResult.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * <p>The canonical example is expiration-related work (either auto-renewal or auto-removal of an
+ * entity).
+ */
+public enum SystemTaskResult {
+    /** Either the id did not refer to an existing entity; or there was no work to for the task. */
+    NOTHING_TO_DO,
+    /**
+     * The id referred to an entity with work to be done for this task, but the work could not be
+     * completed in the current context.
+     */
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskResult.java:30]: Currently `NEEDS_DIFFERENT_CONTEXT` is returned only when the preceding and following child records exceed 54 ?

File: hedera-node/src/main/java/com/hedera/services/stream/RecordStreamObject.java [hunk 6]
```diff
@@ -137,6 +137,7 @@ public String toString() {
                 .append("TransactionRecord", transactionRecord)
                 .append("Transaction", transaction)
                 .append("ConsensusTimestamp", consensusTimestamp)
+                .append("Sidecars", sidecars)
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/stream/RecordStreamObject.java:140]: Is there any reason to not add sidecars to equals and hashCode as well ?

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java [hunk 7]
```diff
@@ -368,24 +382,35 @@ public void deserialize(final SerializableDataInputStream in, final int version)
...
-                                    (lastExpiryUsage == null) ? null : lastExpiryUsage.toJava());
-                }
+                readPreBlockData(in, version);
             }
             blockHashes.clear();
             in.readSerializable(true, () -> blockHashes);
         }
     }
 
+    private void readPreBlockData(final SerializableDataInputStream in, final int version)
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java:392]: nit: method name doesn't really suggest the content logic which includes checks for version 0.30 and 0.31.
Q: Is this supposed to be reusable across upgrades e.g. in future would you want this to do the same in 0.34 to 0.35?

**Review by Nana-EC — APPROVED** [review]
I really like the separation work done here. 
One minor comment around the `readPreBlockData()` method

**Review by Neeharika-Sompalli — DISMISSED** [review]
LGTM ! Just some questions and nits

File: hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryRecordsHelper.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/expiry/ExpiryRecordsHelper.java:117]: 💯 Will add in touch-up PR.

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java:114]: 💯

File: hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java (hunk 4 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java:90]: Sonar is down right now, you're probably right 😁

File: hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskResult.java (hunk 5 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskResult.java:30]: Correct!

File: hedera-node/src/main/java/com/hedera/services/stream/RecordStreamObject.java (hunk 6 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/stream/RecordStreamObject.java:140]: Will do in the cleanup PR. 😁

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java (hunk 7 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java:392]: You caught me, this was just a hack to reduce the cognitive complexity of the method for Sonar. 😓 

Will ponder a cleaner refactoring. 🤔

**Pull request merged**